overseer/src/lib.rs

```diff
@@ -0,0 +1,623 @@
+///
+/// These are:
+///   * Channels being closed
+///   * Subsystems dying when they are not expected to
+///   * Subsystems not dying when they are told to die
+///   * etc.
+// TODO: populate with actual error cases.
+#[derive(Debug)]
+pub struct SubsystemError;
+
+/// A `Result` type that wraps `SubsystemError` and an empty type on success.
+// TODO: Proper success type.
+pub type SubsystemResult = Result<(), SubsystemError>;
+
+/// An asynchronous job that runs inside and being overseen by the `Overseer`.
+///
+/// In essence it's just a newtype wrapping a pinned `Future` dyn trait object.
+pub struct SubsystemJob(pub Pin<Box<dyn Future<Output = ()> + Send + 'static>>);
+
+/// A type of messages that are used inside the `Overseer`.
+///
+/// It is generic over some `M` that is intended to be a message type
+/// being used by the subsystems running on the `Overseer`. Most likely
+/// this type will be one large `enum` covering all possible messages in
+/// the system.
+/// It is also generic over `I` that is entended to be a type identifying
+/// different subsystems, again most likely this is a one large `enum`
+/// covering all possible subsystem kinds.
+enum OverseerMessage<M: Debug, I> {
+	/// This is a message generated by a `Subsystem`.
+	/// Wraps the messge itself and has an optional `to` of
+	/// someone who can receive this message.
+	///
+	/// If that `to` is present the message will be targetedly sent to the intended
+	/// receiver. The most obvious use case of this is communicating with children.
+	SubsystemMessage {
+		to: Option<I>,
+		msg: M,
+	},
+	/// A message that wraps something the `Subsystem` is desiring to
+	/// spawn on the overseer and a `oneshot::Sender` to signal the result
+	/// of the spawn.
+	SpawnChild {
+		s: (I, Box<dyn Subsystem<M, I> + Send>),
+		res: oneshot::Sender<Result<I, SubsystemError>>,
+	},
+}
+
+impl<M: Debug, I: Debug> Debug for OverseerMessage<M, I> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			OverseerMessage::SubsystemMessage { to, msg } => {
+				write!(f, "OverseerMessage::SubsystemMessage{{ to: {:?}, msg: {:?} }}", to, msg)
+			}
+			OverseerMessage::SpawnChild { .. } => write!(f, "OverseerMessage::Spawn(..)")
+		}
+	}
+}
+
+/// A running instance of some `Subsystem`.
+struct SubsystemInstance<M: Debug, I> {
+	/// We talk to the `Overseer` over this channel.
+	rx: mpsc::Receiver<OverseerMessage<M, I>>,
+	/// The `Overseer` talks to use over this channel.
+	tx: mpsc::Sender<M>,
+}
+
+/// An `id` that is given to any `SubsystemInstance` for identification.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct SubsystemId(usize);
+
+/// A context type that is given to the `Subsystem` upon spawning
+/// that can be used by `Subsystem` to communicate with the outside world.
+pub struct SubsystemContext<M: Debug, I>{
+	rx: mpsc::Receiver<M>,
+	tx: mpsc::Sender<OverseerMessage<M, I>>,
```

Comment:
Instead of having a `Subsystem` send messages via a channel, why not have it implement `Stream` and return these messages on `poll_next`? I would find the resulting API easier to understand, there are less sideeffects and it is close to the futures ecosystem standards (using the `Stream` trade for something that produces values).

The `FuturesUnordered` could be replaced by [`StreamsUnordered`](https://github.com/jonhoo/streamunordered).

Reply:
A couple of reasons:
1. I would like to avoid manual implementations of `Stream`s and `Futures`
2. Generally speaking it's quite hard to turn a very abstract logic inside the `async` blocks that we are trying to allow the user implement into a `Stream` without using some intermediary underneath as `mpsc`, at least idk of any at the moment. Do you? Probably `yeild` could do that but I haven't looked closely

Reply:
Channels should be fine, and as Fedor says it should be easier to implement futures that send messages on channels than things that yield messages at the moment. Given that this PR blocks quite a lot of work, IMO we should not block this PR on any refactoring to use streams, although I am not opposed to such refactoring in the future.